### PR TITLE
Scrollbar completion options 

### PR DESCRIPTION
**Pull request by neeasade**
rel issue: https://github.com/The-Compiler/qutebrowser/issues/117

EDIT: I just realized I need to change the test still (tests/unit/config/test_style.py).

### CODE REVIEW

**Comment by The-Compiler** [comment]
As you mentioned the test for the re-added `config` attribute for stylesheets is missing - let me know if you need help with that!

Also I noticed there's no scrollbar displayed at all when there are too many entries:

scrollbar

I think that's quite confusing, but I'm not sure if there's anything you can do...

File: qutebrowser/completion/completionwidget.py [hunk 1]
```diff
@@ -81,10 +81,28 @@ class CompletionView(QTreeView):
...
+
+        QTreeView QScrollBar::handle {
+            /*
+            This is done this way to evade 'magic'
+            ref: http://git.io/vCMqZ
+            Will need to be changed when we utilize jinja better.
+            */
+            background: {{ config.get('colors', 'completion.scrollbar.fg') }};
+            border: {{ config.get('completion', 'scrollbar.padding') }}px solid
+                    {{ config.get('colors', 'completion.scrollbar.bg') }};
```
Comment (The-Compiler) [inline qutebrowser/completion/completionwidget.py]: Adding something like `min-height: 10px;` seems to help with the issue I mentioned in the comment.

Comment (The-Compiler) [inline qutebrowser/completion/completionwidget.py]: Also please dedent this line so it starts where `border:` starts - it's a completely new line, after all. You also don't need the `;` at the end. :smile:

Comment (neeasade) [inline qutebrowser/completion/completionwidget.py]: >Adding something like min-height: 10px; seems to help with the issue I mentioned in the comment.

Ah thanks, will add that.

> Also please dedent this line so it starts where border: starts - it's a completely new line, after all. You also don't need the ; at the end. :smile:

That line is part of the border css. eg with padding of 2 and scrollbar.bg of #000000 that line would turn to `border: 2px solid #000000`

**Comment by neeasade** [comment]
> As you mentioned the test for the re-added config attribute for stylesheets is missing - let me know if you need help with that!

I was not very sure here - never used pytest - I'm guessing I would need to make a test for a dummy objreg.get() call and ensure the value is not changed? My initial approach was to see if there was an existing test that got removed when config did but I did not spot one.

**Comment by The-Compiler** [comment]
Looking at [this line](https://github.com/The-Compiler/qutebrowser/blob/master/tests/unit/config/test_style.py#L34) I think I was in the middle of adding a test and then thought I don't actually need the feature and removed it :laughing:

I think you can just adjust `test_get_stylesheet` there to add something like `{{ config.get('foo', 'bar') }}` to the template and make sure the right thing comes out. I think it'd be better to use `@pytest.mark.parametrize` to split this test into three though, and maybe add some more test cases - something like this:

```python
@pytest.mark.parametrize('template, expected', [
    ("{{ color['completion.bg'] }}", "background-color: black;"),
    ("{{ color['empty.bg'] }}", ""),
    ("{{ color['completion.fg'] }}", "color: red;"),
    ("{{ color['empty.fg'] }}", ""),
    ("{{ font['completion'] }}", "font: foo;"),
    ("{{ font['empty'] }}", ""),
    ("{{ config.get('foo', 'bar') }}", "baz"),
])
def test_get_stylesheet(config_stub, template, expected):
    config_stub.data = {
        'colors': {
            'completion.bg': 'black',
            'completion.fg': 'red',
            'empty.fg': None,
            'empty.bg': None,
        },
        'fonts': {
            'completion': 'foo',
            'empty': None,
        },
        'foo': {'bar': 'baz'},
    }
    rendered = style.get_stylesheet(template)
    assert rendered == expected
```

Could you try if that works for me and change the test to that if it does?

You can run the relevant tests with something like `tox -e py34 -- tests/unit/config/test_style.py` after installing tox.

File: qutebrowser/completion/completionwidget.py (hunk 1 above)
Comment (The-Compiler) [inline qutebrowser/completion/completionwidget.py]: Oh, you're right - sorry.

**Comment by neeasade** [comment]
Ah, I think I understand - so changing the `def test_get_stylesheet` to match the above and test the template resolution on that stub. 

The 'empty' tests above all fail with both None and '' values, though - 
``` pytest
        rendered = style.get_stylesheet(template)
>       assert rendered == expected
E       assert 'color: ;' == ''
E         - color: ;
```

**Comment by The-Compiler** [comment]
Oh. Now I'm confused - to my defense, we're talking about code originally added in one of the first 150ish commits - 9bf7dd2bbb56e9a91eba134fea54550a9c6c1338 :wink:

Just remove the tests with the empty values for now - this really needs a refactoring after that :smile:

**Comment by neeasade** [comment]
All good, I appreciate the guidance.

**Comment by neeasade** [comment]
https://github.com/neeasade/qutebrowser/commit/9b5c0075b96a2b9452009cd57248e9a16e288ca3 Changes the color calls to work with https://github.com/The-Compiler/qutebrowser/pull/1021.

**Comment by neeasade** [comment]
closing, resolved conflicts here: https://github.com/The-Compiler/qutebrowser/pull/1023

**Pull request closed without merging**